### PR TITLE
fix(provisioning): make the Virtualmin pipeline actually provision (#325)

### PR DESCRIPTION
**Pull request by mostlyvirtual**
Closes #325. Closes #263 (the invalid `retry=` kwarg finding is fixed here with a pinning test). Cut from `master`; independent of the other open PRs.

## The defects

The Virtualmin pipeline could not complete its core paths — not edge cases, the happy paths:

1. **Every queued auto-provisioning task crashed on dequeue.** `provision_virtualmin_account_async` passed `retry=TASK_MAX_RETRIES`, which django-q2 1.9.0 does not accept as an option; it leaked into the task kwargs, so the worker called `provision_virtualmin_account(params, retry=3)` → `TypeError`, every time (confirms #263 as always-fails). ADR-0019's own enqueue sample has no `retry` kwarg.
2. **Every account creation failed at pre-flight.** `_validate_provisioning_preconditions` called `self.health_check_server(...)` — a method on a *different* service class — masked by a `# type: ignore[attr-defined]` and swallowed into `Err("Validation error: ...")`. Behind it, the conflict/template checks read response keys the parser never produces, so they always iterated empty lists.
3. **The retry safety net stranded jobs forever.** The 15-minute sweep re-dispatched `create_domain` with an obsolete 3-positional signature; even with the right signature, a fresh create is a structural no-op (the account row survives failure with `status="error"`, so the existing-account check early-exits) while the job — already flipped to `pending` — left the `status="failed"` sweep filter permanently.
4. **Suspension never reached Virtualmin.** Dunning (grace expiry, metering), order cancellation/refund, staff and group actions all flipped `Service.status` while the customer's hosting stayed live; the suspend/unsuspend task chain existed end-to-end with **zero production callers**. Reactivation was separately absorbed by the provisioning trigger's existing-account early-exit. The trigger itself lived inside the *audit* signal handler, gated by `DISABLE_AUDIT_SIGNALS` and killed by any audit failure.
5. **Drift logging crashed exactly when drift existed.** Both `VirtualminDriftRecord` writes used field names that don't exist on the model (`account=`, `drift_description=`, `resolution_action=`), hidden by `# type: ignore[misc]`.
6. **Placement was starved ~50 minutes of every hour.** `is_healthy` demanded a check fresher than 600s while the sweep ran hourly; one failed check set `status="failed"`, removing the server from placement *and* from the sweep's own queryset — permanent eviction with no recovery — while stamping `last_health_check` as if verified.

## The fix, by commit

- **Schema groundwork** — `claimed_at` lease marker on jobs; `consecutive_health_failures` + `failed_by_health_check` on servers (existing failed rows deliberately stay operator-owned: `health_check_error` is not reliable provenance).
- **Enqueue contract + leased job retry** — `retry=` dropped (both sites). The sweep claims each due job with a conditional update (attempt consumed at claim, so broken dispatch can never rearm the same attempt forever) and hands it a job-aware `retry_virtualmin_job` that recovers on the **existing** account+job rows. Execution is fenced (`pending→running` CAS — duplicates self-discard); expired claims (pending *or* running — worker death) return to the failed pool; enqueue failure restores with backoff; exhausted jobs opt out (`next_retry_at=None`); the dispatched task id is persisted. All job-state conditional updates are model classmethods, so no lint-fsm exemption was needed.
- **Convergent create retry with ownership proof** — a timed-out-but-remotely-successful create is finalized in place (same credentials, exactly-once domain-count increment) only after a multiline ownership probe confirms the remote `Username:` is ours; a different owner is a terminal conflict (never adopt another tenant's domain); indeterminate ownership and probe errors are retriable, never treated as absence.
- **Pre-flight that actually passes** — connectivity via the already-supplied gateway; conflicts via the owner-carrying normalized listing (detects both domain and username collisions); templates via a new normalized `list_templates()`; the stateful mock extended so tests run through the real parser. First-ever create-path tests: happy path, connection-failure overshoot guard, domain/username conflicts, missing template.
- **Drift records use real fields**, and the enforce path records `status_mismatch`/`auto_fixed` instead of an out-of-vocabulary drift type.
- **Health model** — 10-minute sweep (Schedule row **upserted** by name: skip-if-exists left deployed installs hourly forever), freshness 1500s, success-only `last_health_check`, `is_healthy` also requires no live error; ≥6 consecutive failures auto-fail reversibly (`failed_by_health_check`), the sweep keeps probing those (the gateway's server-status gate now admits them — previously the recovery branch was unreachable in production), a later success auto-recovers via conditional updates so an operator's concurrent reclassification always wins; `healthy: false` payloads count as failures.
- **Reconcile, don't command** — a dedicated `post_save` handler (separate from the audit handler, indifferent to `DISABLE_AUDIT_SIGNALS`) enqueues an idempotent reconcile task via `transaction.on_commit` (suspension callers hold open atomics; a rolled-back transition never dispatches). The task reads the *committed* Service+account state and converges: active+no account → auto-provision (gated by the ADR-0019 kill switch `VIRTUALMIN_AUTO_PROVISIONING_ENABLED`, implemented at last); active+suspended → unsuspend; suspended/terminated/expired+active → suspend (never delete). It re-reads the Service after acting and re-queues once if the state moved mid-flight; lifecycle retries verify current desired state first, so an old unsuspend job can never re-enable a since-suspended customer. A post-create reconcile closes the termination-during-create window.
- **Idempotency self-heal** — cached suspend/unsuspend successes are honored only when the account actually matches the target state; flip-flops inside the 1-hour TTL execute for real; error paths clear the in-progress sentinel instead of absorbing reconciliation for an hour.

## Judgement calls

- **Existing `failed` servers stay operator-owned** (no migration backfill): auto-reactivating a deliberately-failed server would be worse than one manual reactivation per legacy row.
- **Terminated services suspend, never delete** — deletion stays protected/manual per ADR-0019.
- **Reject-shaped tests were inverted deliberately**: the old single-strike health test asserted the defect (one failure → `failed` + stamped timestamp) and now pins the new contract.

## Verification

- 41 tests in the previously-zero-coverage task module (all written red-first against the actual defects) + inverted/updated service tests; full provisioning package green; full platform fast suite green; per-file lint and mypy clean on every touched module; `make lint-fsm` clean with no new exemptions.
- Plan was gated twice (internal reviewer + two external review passes) before implementation; the implementation then survived a four-lens adversarial review + independent conformance check, whose findings (ownership adoption, unreachable auto-recovery, claim fencing, snapshot-race reconciliation, idempotency absorption) are all fixed and regression-pinned in the final commit.

Known limitations (deliberate): the mock's `list_templates` does not exercise the production parser end-to-end; the cluster-wide `Q_CLUSTER retry(600) < per-task timeout(900)` tension is pre-existing and tracked with #328's task-layer items.

### CODE REVIEW

File: services/platform/apps/provisioning/virtualmin_service.py [hunk 1]
```diff
@@ -379,8 +379,10 @@ def _validate_provisioning_preconditions(
...
         """
         try:
-            # 1. Server health check
-            health_result = self.health_check_server(account.server)  # type: ignore[attr-defined]
+            # 1. Server connectivity (the gateway is already bound to the
+            # target server; no status side effects, unlike the management
+            # service's health_check_server)
+            health_result = gateway.test_connection()
             if health_result.is_err():
                 return Err(f"Server health check failed: {health_result.unwrap_err()}")
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/virtualmin_service.py:387]: Addressed in 329d70d5.

File: services/platform/apps/provisioning/signals.py [hunk 2]
```diff
@@ -325,14 +325,49 @@ def audit_service_lifecycle_events(sender: type[Service], instance: Service, cre
...
+    Reconcile-not-command: the task reads the COMMITTED Service+account
+    state and converges Virtualmin to it, so stale commands and
+    multi-transition ordering races cannot occur; duplicate enqueues are
+    harmless. Enqueue happens on_commit — suspension/reactivation callers
+    hold open atomics, and work for a rolled-back transition must never run.
+    """
+    update_fields = kwargs.get("update_fields")
+    status_changed = created or (update_fields and "status" in update_fields)
+    if not status_changed:
+        return
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/signals.py:359]: Addressed in 329d70d5.

File: services/platform/apps/provisioning/virtualmin_service.py [hunk 3]
```diff
@@ -1177,24 +1337,57 @@ def health_check_server(self, server: VirtualminServer) -> Result[dict[str, Any]
...
-                server.status = "failed"
-                server.save(update_fields=["last_health_check", "health_check_error", "status", "updated_at"])
+                # Failure: record the error and the streak. last_health_check
+                # is NOT stamped (it means "last verified"), and a single
+                # strike no longer evicts the server — only a sustained
+                # streak auto-fails it, reversibly.
+                VirtualminServer.objects.filter(pk=server.pk).update(
+                    health_check_error=error_msg,
+                    consecutive_health_failures=models.F("consecutive_health_failures") + 1,
+                )
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/virtualmin_service.py:1418]: Addressed in 329d70d5.

File: services/platform/apps/provisioning/virtualmin_service.py [hunk 4]
```diff
@@ -1177,24 +1337,57 @@ def health_check_server(self, server: VirtualminServer) -> Result[dict[str, Any]
...
+                server.refresh_from_db()
+                if server.status == "active" and server.consecutive_health_failures >= HEALTH_AUTO_FAIL_THRESHOLD:
+                    VirtualminServer.objects.filter(pk=server.pk, status="active").update(
+                        status="failed", failed_by_health_check=True
+                    )
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/virtualmin_service.py]: Addressed in 329d70d5.

File: services/platform/apps/provisioning/virtualmin_models.py [hunk 5]
```diff
@@ -630,6 +647,49 @@ class Meta:
...
+    def claim_for_retry(cls, job_id: Any, now: Any) -> bool:
+        """Leased claim: failed -> pending, consuming one attempt atomically."""
+        return bool(
+            cls.objects.filter(
+                pk=job_id,
+                status="failed",
+                retry_count__lt=models.F("max_retries"),
+                next_retry_at__lte=now,
+            ).update(status="pending", retry_count=models.F("retry_count") + 1, claimed_at=now)
+        )
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/virtualmin_models.py:660]: Addressed in 329d70d5.

File: services/platform/apps/provisioning/virtualmin_models.py [hunk 6]
```diff
@@ -630,6 +647,49 @@ class Meta:
...
+
+    @classmethod
+    def start_claimed(cls, job_id: Any, now: Any) -> bool:
+        """Fence at execution: pending -> running. 0 rows = stale duplicate delivery."""
+        return bool(cls.objects.filter(pk=job_id, status="pending").update(status="running", started_at=now))
+
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/virtualmin_models.py]: Addressed in 329d70d5.

File: services/platform/apps/provisioning/virtualmin_models.py [hunk 7]
```diff
@@ -630,6 +647,49 @@ class Meta:
...
+    @classmethod
+    def recover_expired_claims(cls, cutoff: Any, retry_at: Any) -> int:
+        """Claimed jobs (pending OR running) whose lease expired go back to failed."""
+        return cls.objects.filter(
+            models.Q(status="pending") | models.Q(status="running"),
+            claimed_at__isnull=False,
+            claimed_at__lt=cutoff,
+        ).update(status="failed", next_retry_at=retry_at)
+
```
[the remark this answers is not in the document]
Comment (mostlyvirtual) [reply services/platform/apps/provisioning/virtualmin_models.py:698]: Addressed in 329d70d5.
